### PR TITLE
[metal] Initial bring up of Metal HAL driver (6/n)

### PR DESCRIPTION
**Pull request by antiagainst**
This commit works on implementing hal::DebugCaptureManager.
We just wrap a MTLCaptureManager under the hood.

### CODE REVIEW

File: iree/hal/metal/metal_driver.mm [hunk 1]
```diff
@@ -16,10 +16,14 @@
 
 #import <Metal/Metal.h>
 
+#include "absl/flags/flag.h"
 #include "iree/base/status.h"
 #include "iree/base/tracing.h"
+#include "iree/hal/metal/metal_capture_manager.h"
 #include "iree/hal/metal/metal_device.h"
 
+ABSL_FLAG(bool, metal_capture, false, "Enables capturing Metal commands.");
```
Comment (benvanik) [inline iree/hal/metal/metal_driver.mm:25]: we should be keeping flags out of these files and in the _driver_module.cc files instead (it was a mistake that the renderdoc flag ended up in vulkan_driver - that should be moved out of there and passed in via the VulkanDriver::Options struct instead).

this is important for size-sensitive builds as absl::Flags is like >100KB of binary size increase and we want to be able to compile it out.

File: iree/hal/metal/metal_capture_manager.h [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+#include "iree/base/status.h"
+#include "iree/hal/debug_capture_manager.h"
+
+namespace iree {
+namespace hal {
+namespace metal {
+
+// A DebugCaptureManager implementation for Metal that directly wraps a
+// MTLCpatureManager.
```
Comment (ScottTodd) [inline iree/hal/metal/metal_capture_manager.h:30]:
```suggestion
// MTLCaptureManager.
```
typo

File: iree/hal/metal/metal_capture_manager.mm [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+#include <string>
+
+#include "absl/flags/flag.h"
+#include "iree/base/file_io.h"
+#include "iree/base/logging.h"
+#include "iree/base/tracing.h"
+
+ABSL_FLAG(std::string, metal_capture_to_file, "",
+          "Full path to store the GPU trace file (empty means capture to Xcode)");
```
Comment (ScottTodd) [inline iree/hal/metal/metal_capture_manager.mm:25]: can also move this flag

File: iree/hal/metal/metal_driver.mm (hunk 1 above)
Comment (antiagainst) [reply iree/hal/metal/metal_driver.mm]: Good point. Makes sense. Moved.

**Review by ScottTodd — APPROVED** [review]
nit: these PR / commit titles aren't descriptive:
https://github.com/google/iree/tree/main/iree/hal/metal

image

(this one in particular is mainly MetalCaptureManager)

**Comment by antiagainst** [comment]
They are consistent. ;-P 

Modified the title a bit to be descriptive.

**Pull request merged**